### PR TITLE
Initial KeyVaultProxy solution

### PR DESCRIPTION
**Pull request by heaths**
Ported with permission from @heaths from https://github.com/heaths/KeyVaultProxy. Builds both within and without the Azure/azure-sdk-for-net repo.

### CODE REVIEW

**Comment by heaths** [comment]
I'm moving this from my person repo where I used it as an example for both the KV service team and on social like Stack Overflow. It builds in and out of the repo correctly.

I am considering, though, that maybe the README with the sample front matter should be only in the _src_ directory, since:

1. The tests really aren't needed to be part of the sample, right?
2. I can more easily consume the existing test framework we have already for live tests - whether we run them nightly or not - rather than duplicate a few TestFramework sources and use my xunit-based test discovery.

Do people agree I should just make the _src_ project itself the published sample and leave the tests here? Should I make sure they run as part of nightlies?

File: sdk/keyvault/Microsoft.Azure.KeyVault/tests/Microsoft.Azure.KeyVault.Tests.csproj [hunk 1]
```diff
@@ -27,6 +27,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" />
```
Comment (weshaggard) [inline sdk/keyvault/Microsoft.Azure.KeyVault/tests/Microsoft.Azure.KeyVault.Tests.csproj:30]: What is this needed?

**Review by weshaggard — APPROVED** [review]
I didn't look deeply at the code but the general structure looks fine. I assume you add the snippets to make sure they stay updated in the MD.

**Comment by weshaggard** [comment]
>Do people agree I should just make the src project itself the published sample and leave the tests here? Should I make sure they run as part of nightlies?

I'm Ok with having the tests here, but I'm not sure we should add them to our nightly test runs. Eventually we should get the sample code running in our smoke tests and given this is kind of a sample library we should either write a sample application or maybe use the tests as a proxy and run as part of those smoke tests.

File: sdk/keyvault/Microsoft.Azure.KeyVault/tests/Microsoft.Azure.KeyVault.Tests.csproj (hunk 1 above)
Comment (heaths) [reply sdk/keyvault/Microsoft.Azure.KeyVault/tests/Microsoft.Azure.KeyVault.Tests.csproj:30]: Yes, for the callback that `KeyVaultClient` needs. Any concern?

**Comment by heaths** [comment]
> I'm Ok with having the tests here, but I'm not sure we should add them to our nightly test runs.

I agree keeping the test source is good, but I was referring more to whether only `src` should be published as the sample (i.e. the sample README should go there). The more I think about it, the more I think that's probably a better call. The tests and structure to support the tests only unnecessarily complicate the sample.

As for getting them running, I'll open a test issue on the backlog. I had considered this, but even getting sample source that builds and runs correctly in our repo **and** as a standalone downloaded sample is enough of a pain without complicating the source. I brought this up yesterday in our .NET team meeting as well (other people are looking to add samples) and want to with you more about it.

File: sdk/keyvault/Microsoft.Azure.KeyVault/tests/Microsoft.Azure.KeyVault.Tests.csproj (hunk 1 above)
Comment (weshaggard) [reply sdk/keyvault/Microsoft.Azure.KeyVault/tests/Microsoft.Azure.KeyVault.Tests.csproj:30]: This is track 1 and a test project so I'm not concerned with this. We do need to be a little careful about introducing dependencies though.
